pkg/order/syncer/state_syncer.go

```diff
@@ -0,0 +1,323 @@
+		return err
+	}
+
+	var parentBlockHash *types.Hash
+	for i, rangeHeight := range rangeHeights {
+		if i == 0 {
+			parentBlockHash = metaHash
+		}
+		rangeTmp := rangeHeight
+		headers := s.syncQuorumRangeBlockHeaders(rangeTmp, parentBlockHash)
+		if headers == nil {
+			return fmt.Errorf("fetch and verify the quorum peers' block header error: %v", rangeTmp)
+		}
+		blocks := s.syncRangeBlocks(headers)
+		if blocks == nil {
+			return fmt.Errorf("fetch and verify peers' block error: %v", rangeTmp)
+		}
+		for _, block := range blocks {
+			blockCh <- block
+		}
+	}
+
+	return nil
+}
+
+func (s *StateSyncer) syncQuorumRangeBlockHeaders(rangeHeight *rangeHeight, parentBlockHash *types.Hash) []*pb.BlockHeader {
+	var isQuorum bool
+	var hash *types.Hash
+	latestBlockHeaderCounter := make(map[*types.Hash]uint64)
+	blockHeadersM := make(map[*types.Hash][]*pb.BlockHeader)
+
+	fetchAndVerifyBlockHeaders := func(id uint64) {
+		s.logger.WithFields(logrus.Fields{
+			"begin":   rangeHeight.begin,
+			"end":     rangeHeight.end,
+			"peer_id": id,
+		}).Info("syncing range block header")
+		headers, err := s.fetchBlockHeaders(id, rangeHeight.begin, rangeHeight.end)
+		if err != nil {
+			s.logger.Errorf("fetch block headers error:%w", err)
+			return
+		}
+		err = s.verifyBlockHeaders(parentBlockHash, headers)
+		if err != nil {
+			s.logger.Errorf("check block headers error:%w", err)
+			return
+		}
+		latestBlock := &pb.Block{BlockHeader: headers[len(headers)-1]}
+		blockHash := latestBlock.Hash()
+		latestBlockHeaderCounter[blockHash]++
+		blockHeadersM[blockHash] = headers
+	}
+
+	for _, id := range s.peerIds {
+		for latestHash, counter := range latestBlockHeaderCounter {
+			if counter >= s.quorum {
+				hash = latestHash
+				isQuorum = true
+				break
+			}
+		}
+		if isQuorum {
+			break
+		}
+		fetchAndVerifyBlockHeaders(id)
+	}
+	if hash == nil {
+		return nil
+	}
+
+	return blockHeadersM[hash]
+
+}
+
+func (s *StateSyncer) syncRangeBlocks(headers []*pb.BlockHeader) []*pb.Block {
+	var blocks []*pb.Block
+	begin := headers[0].Number
+	end := headers[len(headers)-1].Number
+
+	fetchAndVerifyBlocks := func(id uint64) {
+		s.logger.WithFields(logrus.Fields{
+			"begin":   begin,
+			"end":     end,
+			"peer_id": id,
+		}).Info("syncing range block")
+		fetchBlocks, err := s.fetchBlocks(id, begin, end)
+		if err != nil {
+			s.logger.Errorf("fetch block headers error:%w", err)
+			return
+		}
+		for i, block := range fetchBlocks {
+			err := s.verifyBlock(headers[i], block)
+			if err != nil {
+				s.logger.Errorf("check block headers error:%w", err)
+				return
+			}
+		}
+		blocks = fetchBlocks
+	}
+	for _, id := range s.peerIds {
+		if blocks != nil {
+			break
+		}
+		fetchAndVerifyBlocks(id)
+	}
+	return blocks
+}
+
+func (s *StateSyncer) randPeers() uint64 {
+	ids := make([]uint64, 0)
+	for _, id := range s.peerIds {
+		_, ok := s.badPeers.Load(id)
+		if ok {
+			continue
+		}
+		ids = append(ids, id)
+	}
+	if len(ids) == 0 {
+		return 0
```

Comment:
cannot return 0 if there is no peers, otherwise it may be considered as peer id 0